codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/PaginatorGenerator.kt

```diff
@@ -211,9 +223,37 @@ class PaginatorGenerator private constructor(
                             "RuntimePlugins" to RuntimeType.runtimePlugins(runtimeConfig),
                         )
                     },
+                "is_empty_setter" to isEmptySetter(),
             )
         }
 
+    /** Generate code to calculate the value of is_empty. For most paginators this
+     * is indicated by the next token being the empty string. But for paginators
+     * with the isTruncatedPaginator trait the next token is not necessarily empty.
+     * (ex: for s3 ListParts the final next token is "0" when pagination is complete,
+     * causing the paginator to go back to the first page and loop forever)
+     * In this case we use a false value of isTruncated as the only indicator that
+     * the pagination is exhausted.
+     * */
+    private fun isEmptySetter() =
```

Comment:
If you wanted to go even more down the rabbit hole you could declare a [section](https://github.com/smithy-lang/smithy-rs/blob/fe36727499a7a74e42a2beca7e76d0b433871181/codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/customize/Customization.kt#L30) for this bit of code and then register a customization for that section that does something different. You would then basically be able to move all of this into a single `TruncatedPaginatorDecorator` that handles both the model transform and the actual codegen logic.